unitycatalog-ai/integrations/autogen/README.md

```diff
@@ -0,0 +1,197 @@
+```
+
+#### Create a function in UC
+
+Create a python UDF in Unity Catalog with the client
+
+```python
+# replace with your own catalog and schema
+CATALOG = "catalog"
+SCHEMA = "schema"
+
+func_name = f"{CATALOG}.{SCHEMA}.python_exec"
+# define the function body in SQL
+sql_body = f"""CREATE OR REPLACE FUNCTION {func_name}(location STRING COMMENT 'Retrieves the current weather from a provided location.')
+RETURNS STRING
+LANGUAGE PYTHON
+COMMENT 'Returns the current weather from a given location and returns the temperature in degrees Celsius.'
+AS $$
+    return "31.9 C"
+$$
+"""
+
+client.create_function(sql_function_body=sql_body)
+```
+
+Now that the function is created and stored in the corresponding catalog and schema, we can use it within autogen's SDK.
+
+#### Create a UCFunctionToolkit instance
+
+To begin, we will need an instance of the tool function interface from the `ucai_autogen` toolkit.
+
+```python
+from ucai_autogen.toolkit import UCFunctionToolkit
+
+# Create an instance of the toolkit with the function that was created earlier.
+toolkit = UCFunctionToolkit(function_names=[f"{CATALOG}.{SCHEMA}.python_exec"], client=client)
+
+# Access the tool definitions that are in the interface that autogen's SDK expects
+tools = tookit.tools
+
+```
+
+Now that we have the defined tools from Unity Catalog, we can directly pass this definition to the Conversable Agent API within Autogen.
+
+If you would like to validate that your tool is functional prior to proceeding to integrate it with LlamaIndex, you can call the tool directly:
+
+```python
+my_tool = tools[0]
+
+my_tool.fn(**{"location": "San Franscisco"})
+```
+
+#### Use the tools with a conversable Agent
+
+```python
+
+import os
+from autogen import ConversableAgent
+
+
+OPENAI_API_KEY ='.....'
+
+# Let's first define the assistant agent that suggests tool calls.
+assistant = ConversableAgent(
+    name="Assistant",
+    system_message="You are a helpful AI assistant. "
+    "You can tell the temperature of a location "
+    "Return 'TERMINATE' when the task is done.",
+    llm_config={"config_list": [{"model": "gpt-4", "api_key": OPENAI_API_KEY}]},
+)
+
+# The user proxy agent is used for interacting with the assistant agent
+# and executes tool calls.
+user_proxy = ConversableAgent(
+    name="User",
+    llm_config=False,
+    is_termination_msg=lambda msg: msg.get("content") is not None and "TERMINATE" in msg["content"],
+    human_input_mode="NEVER",
+)
+
+```
+
+Once you have created a tool, you can register it with the agents that are involved in the conversation.
+
+Similar to code executors, a tool must be registered with at least two agents for it to be useful in conversation. The agent which can call the tool and and Agent whic can execute the tool’s function.
+
+you can use `register_function` method to register a tool with both agents at once.
+below is an example
+
+```python
+
+for tool in tools:
+    tool.register_function(assistant,user_proxy)
```

Comment:
1. Can we use named arguments here to provide guidance to readers about which args these are setting?
2. Can we update the example to show how to set a collection of tools on separate agents? While intuitively it seems like this would look something like:

```python
agents = [
  {"caller": assistant_1, "executor": user_proxy_1}, 
  {"caller": assistant_2, "executor": user_proxy_2}
  ]
for tool in tools:
  for agent in agents:
    tool.register_function(**agent)
```

This just seems very painful. 

What do you think about a helper method that applies the toolkit's tools to a user-supplied agent?

```python
 def register_with_agents(self, caller: "ConversableAgent", executor: "ConversableAgent") -> None:
        """
        Registers all tools in the toolkit with the specified caller and executor agents.

        Args:
            caller (ConversableAgent): The agent that will call the tools.
            executor (ConversableAgent): The agent that will execute the tools' functions.
        """
        for tool in self.tools:
            tool.register_function(caller=caller, executor=executor)
```

Reply:
Good shout. this gives the flexibility to register all the tools at once as well. I will add the above method and also add that as an example.

Reply:
added a group chat example and documentation about the new method